File: backend/app/models/analysis.py
# 참고: docs/backend/06-company-analysis.md
# 참고: docs/backend/05-llm-service.md

"""
기업 분석 데이터 모델

기업 분석 결과, 재무 분석, 뉴스 분석, 기술적 분석 모델을 정의합니다.
"""

from pydantic import BaseModel, Field
from typing import List, Literal, Optional
from datetime import datetime


OpinionType = Literal["STRONG_BUY", "BUY", "HOLD", "SELL", "STRONG_SELL"]


class FinancialData(BaseModel):
    """재무 데이터"""

    roe: float = Field(..., description="ROE (%)")
    per: float = Field(..., description="PER")
    pbr: float = Field(..., description="PBR")
    debt_ratio: float = Field(..., description="부채비율 (%)")
    revenue_growth_yoy: float = Field(..., description="매출 성장률 (%)")


class NewsData(BaseModel):
    """뉴스 데이터"""

    positive_count: int = Field(..., description="긍정 뉴스 개수")
    negative_count: int = Field(..., description="부정 뉴스 개수")
    neutral_count: int = Field(..., description="중립 뉴스 개수")
    sentiment: Literal["positive", "neutral", "negative"] = Field(..., description="종합 센티먼트")


class TechnicalData(BaseModel):
    """기술적 지표 데이터"""

    rsi: float = Field(..., description="RSI (0-100)")
    macd_status: Literal["golden_cross", "dead_cross", "neutral"] = Field(..., description="MACD 상태")
    ma_position: Literal["상회", "하회", "중립"] = Field(..., description="이동평균선 위치")


class Analysis(BaseModel):
    """기업 분석 결과"""

    # 기본 정보
    ticker: str = Field(..., description="종목 코드")
    name: str = Field(..., description="종목명")
    current_price: int = Field(..., description="현재가 (원)")
    change_rate: float = Field(..., description="등락률 (%)")
    market_cap: int = Field(..., description="시가총액 (억원)")

    # 분석 결과
    summary: str = Field(..., description="한줄 요약")
    opinion: OpinionType = Field(..., description="투자 의견")
    target_price: int = Field(..., description="목표가")

    # 상세 분석
    financial: FinancialData = Field(..., description="재무 분석")
    news: NewsData = Field(..., description="뉴스 분석")
    technical: TechnicalData = Field(..., description="기술적 분석")
    risks: List[str] = Field(..., description="리스크 요인")

    # 메타데이터
    analyzed_at: datetime = Field(default_factory=datetime.now, description="분석 시간")

    class Config:
        json_schema_extra = {
            "example": {
                "ticker": "005930",
                "name": "삼성전자",
                "current_price": 72000,
                "change_rate": 5.2,
                "market_cap": 4_300_000,  # 억원
                "summary": "HBM3 수주 증가로 향후 3개월간 상승 여력 높음",
                "opinion": "STRONG_BUY",
                "target_price": 80000,
                "financial": {
                    "roe": 15.2,
                    "per": 12.5,
                    "pbr": 1.8,
                    "debt_ratio": 45.0,
                    "revenue_growth_yoy": 12.5
                },
                "news": {
                    "positive_count": 8,
                    "negative_count": 2,
                    "neutral_count": 3,
                    "sentiment": "positive"
                },
                "technical": {
                    "rsi": 65.0,
                    "macd_status": "golden_cross",
                    "ma_position": "상회"
                },
                "risks": [
                    "미중 무역 분쟁 심화 가능성",
                    "글로벌 반도체 수요 둔화 우려",
                    "환율 변동성 확대"
                ],
                "analyzed_at": "2025-11-06T10:30:00"
            }
        }


class AnalysisResponse(BaseModel):
    """기업 분석 API 응답"""

    success: bool = Field(..., description="성공 여부")
    data: Analysis = Field(..., description="분석 데이터")

File: frontend/src/types/analysis.ts
// 참고: docs/backend/06-company-analysis.md
// 참고: backend/app/models/analysis.py

/**
 * 기업 분석 관련 타입 정의
 */

export type OpinionType =
  | 'STRONG_BUY'
  | 'BUY'
  | 'HOLD'
  | 'SELL'
  | 'STRONG_SELL'

export interface FinancialData {
  roe: number          // ROE (%)
  per: number          // PER
  pbr: number          // PBR
  debt_ratio: number   // 부채비율 (%)
  revenue_growth: number // 매출 성장률 (%)
}

export interface NewsData {
  positive_count: number
  negative_count: number
  neutral_count: number
  sentiment: 'positive' | 'neutral' | 'negative'
}

export interface TechnicalData {
  rsi: number
  macd_status: 'golden_cross' | 'dead_cross' | 'neutral'
  ma_position: '상회' | '하회' | '중립'
}

export interface Analysis {
  ticker: string
  name: string
  current_price: number
  change_rate: number
  market_cap: number

  // 분석 결과
  summary: string
  opinion: OpinionType
  target_price: number

  // 상세 분석
  financial: FinancialData
  news: NewsData
  technical: TechnicalData
  risks: string[]

  // 메타데이터
  analyzed_at: string
}

export interface AnalysisResponse {
  success: boolean
  data: Analysis
}

export interface MetricScore {
  ticker: string
  name: string
  total_score: number // 0-10
  breakdown: {
    momentum: number
    volume: number
    sentiment: number
    technical: number
    financial: number
  }
  calculated_at: string
}

export interface ComparisonData {
  ticker: string
  name: string
  sector: string
  sector_average: {
    per: number
    pbr: number
    roe: number
  }
  comparison: {
    per: '상회' | '하회' | '유사'
    pbr: '상회' | '하회' | '유사'
    roe: '상회' | '하회' | '유사'
  }
  ranking: {
    market_cap: number
    per: number
    roe: number
  }
  peers: Array<{
    ticker: string
    name: string
    per: number
    pbr: number
  }>
}

// 투자 의견 한글명
export const OPINION_LABELS: Record<OpinionType, string> = {
  STRONG_BUY: '적극 매수',
  BUY: '매수',
  HOLD: '보유',
  SELL: '매도',
  STRONG_SELL: '적극 매도',
}

// 투자 의견별 색상
export const OPINION_COLORS: Record<OpinionType, string> = {
  STRONG_BUY: 'text-red-600 bg-red-50',
  BUY: 'text-red-500 bg-red-50',
  HOLD: 'text-gray-600 bg-gray-50',
  SELL: 'text-blue-500 bg-blue-50',
  STRONG_SELL: 'text-blue-600 bg-blue-50',
}
